### ISSUE TITLE
fqdnHostname contains unqualified hostname on NixOS

### ISSUE THREAD

**Issue by ToxicFrog** [twpayne/chezmoi#3054]
## Describe the bug

The tl;dr version is:
```console
$ hostname; hostname -f
ancilla
ancilla.ancilla.ca
$ chezmoi data | grep -i hostname
    "fqdnHostname": "ancilla",
    "hostname": "ancilla",
```
Digging into it reveals some more information. `/etc/hostname` contains the bare hostname, not the fqdn, and stuff like `uname -n` returns the bare hostname. `/etc/hosts`, however, contains the fqdn:
```console
$ cat /etc/hostname
ancilla
$ uname -n
ancilla
$ cat /etc/hosts
127.0.0.1 localhost
::1 localhost
127.0.0.2 ancilla.ancilla.ca ancilla
::1 ancilla.ancilla.ca ancilla
```
Conveniently, I have a different computer on the same network, running openSUSE instead of NixOS, where fqdnHostname works:

<details>

```console
$ hostname; hostname -f
thoth.ancilla.ca
thoth.ancilla.ca
$ chezmoi data | grep -i hostname
    "fqdnHostname": "thoth.ancilla.ca",
    "hostname": "thoth",
$ cat /etc/hostname
thoth.ancilla.ca
$ uname -n
thoth.ancilla.ca
$ cat /etc/hosts
127.0.0.1       localhost
# fallback hostname used by NetworkManager
127.0.0.1       localhost.localdomain

# special IPv6 addresses
::1             localhost ipv6-localhost ipv6-loopback

fe00::0         ipv6-localnet
ff00::0         ipv6-mcastprefix
ff02::1         ipv6-allnodes
ff02::2         ipv6-allrouters
ff02::3         ipv6-allhosts
```

</details>

The most obvious difference is that on thoth, the fqdn is written to `/etc/hostname` (which, according to the man page, is permitted but not recommended), while on ancilla, it is written to `/etc/hosts` and `/etc/hostname` contains only the unqualified hostname. But the issue is not solely that, because I have a work computer (running Debian) that also has the unqualified hostname there, but where `fqdnHostname` works:

<details>

```console
$ hostname; hostname -f
socius
socius.roam.[redacted]
$ chezmoi data | grep -i hostname
    "fqdnHostname": "socius.roam.[redacted]",
    "hostname": "socius",
$ cat /etc/hostname
socius
$ cat /etc/hosts
127.0.1.1    socius.roam.[redacted] socius
127.0.0.1    localhost
ff02::3 ip6-allhosts
```

</details>

Interestingly, running them in strace reveals different behaviour:
- on ancilla it calls `uname(2)`, then reads `/etc/hosts` and `/etc/hostname`
- on socius it calls `uname` and then reads `/etc/hosts`, but not `/etc/hostname`
- on thoth it also calls `uname` but skips reading those files in `/etc/ entirely

Without looking at the code, my best guess is that on thoth it gets the fqdn from `uname` and skips `/etc/host*` as it knows it doesn't need them; on socius it gets the hostname from `uname` and the fqdn from `/etc/hosts` and stops there; and on ancilla it behaves similarly to socius, but either (a) treats the unqualified hostname in `/etc/hostname` as authoritative for some reason or (b) fails to parse the contents of `/etc/hosts`.

## To reproduce

On a nixos computer:
```console
$ hostname; hostname -f
$ nix-shell -p '(import <nixos-unstable> {}).chezmoi' --run 'chezmoi data' | grep -i hostname
```
and compare the outputs.

## Expected behavior

`fqdnHostname` is `ancilla.ancilla.ca`.

## Output of command with the `--verbose` flag

`chezmoi data` does not log anything extra with `--verbose`.

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
RESULT    CHECK                MESSAGE
warning   version              v2.33.6, built by nixpkgs
warning   latest-version       v2.34.2
ok        os-arch              linux/amd64 (NixOS 22.11 (Raccoon))
ok        uname                Linux ancilla 5.15.111 #1-NixOS SMP Thu May 11 14:00:40 UTC 2023 x86_64 GNU/Linux
ok        go-version           go1.20.4 (gc)
ok        executable           /nix/store/abnsggq8fa42z6nkpjjvlcvq3gmrj6wj-chezmoi-2.33.6/bin/chezmoi
ok        upgrade-method       sudo-upgrade-package
ok        config-file          no config file found
ok        source-dir           ~/.local/share/chezmoi is a git working tree (clean)
ok        suspicious-entries   no suspicious entries
ok        working-tree         ~/.local/share/chezmoi is a git working tree (clean)
ok        dest-dir             ~ is a directory
ok        umask                022
ok        cd-command           found /run/current-system/sw/bin/zsh
ok        cd-args              /run/current-system/sw/bin/zsh
info      diff-command         not set
ok        edit-command         found /run/current-system/sw/bin/nano
ok        edit-args            nano --softwrap -a -w
ok        git-command          found /run/current-system/sw/bin/git, version 2.38.5
warning   merge-command        vimdiff not found in $PATH
ok        shell-command        found /run/current-system/sw/bin/zsh
ok        shell-args           /run/current-system/sw/bin/zsh
info      age-command          age not found in $PATH
info      gpg-command          gpg not found in $PATH
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      dashlane-command     dcli not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keepassxc-db         not set
info      keeper-command       keeper not found in $PATH
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      passhole-command     ph not found in $PATH
info      rbw-command          rbw not found in $PATH
info      vault-command        vault not found in $PATH
info      secret-command       not set
```

</details>

## Additional context

I initially thought this might be a NixOS bug, but according to `man 5 hostname` putting the unqualified domain in `/etc/hostname` is the preferred behaviour, and my work machine does the same thing (and it works there).

I also thought this might be a regression of #1269, but per `strace` it is checking `/etc/hosts`, it's just not successfully extracting the fqdn from it.

**Comment by twpayne** [comment]
Note that man page for `hosts(5)` states

> 127.0.1.1 is often used for the FQDN of the machine

so NixOS using 127.0.0.2 here is a bit weird and might confuse other software running on NixOS.

**Comment by ToxicFrog** [comment]
Thank you for the quick fix!

I'm curious about the genesis of that line in `hosts(5)`; using 127.0.1.1 for the fqdn seems to be common on Debian-derived systems, but not Red Hat/Fedora or SUSE. Not sure how widespread it is in other distros and I don't quite care enough to fire up a bunch of VMs and check.
